File: problem/Smallest_Subarrays_OR/smallestSubarrays.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *smallestSubarrays(int *nums, int numsSize, int *returnSize)
{
    int n = numsSize;
    *returnSize = n;
    int *res = (int *)malloc(n * sizeof(int));
    int *suf = (int *)malloc(n * sizeof(int));
    suf[n - 1] = nums[n - 1];
    for (int i = n - 2; i >= 0; i--)
    {
        suf[i] = suf[i + 1] | nums[i];
    }
    int l = 0, r = 0;
    int cur = nums[0];
    int freq[32] = {0};
    for (int j = 0; j < 32; j++)
    {
        if ((nums[0] & (1U << j)) != 0)
        {
            freq[j]++;
        }
    }
    while (l < n)
    {
        while (r + 1 < n && cur != suf[l])
        {
            r++;
            cur |= nums[r];
            for (int j = 0; j < 32; j++)
            {
                if ((nums[r] & (1U << j)) != 0)
                {
                    freq[j]++;
                }
            }
        }
        res[l] = r - l + 1 > 1 ? r - l + 1 : 1;
        for (int j = 0; j < 32; j++)
        {
            if ((nums[l] & (1U << j)) != 0)
            {
                freq[j]--;
                if (freq[j] == 0)
                {
                    cur = cur & (~(1U << j));
                }
            }
        }
        l++;
    }
    free(suf);
    return res;
}

File: problem/Smallest_Subarrays_OR/smallestSubarrays.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var smallestSubarrays = function(nums) {
    let n = nums.length;
    let res = new Array(n);
    let suf = new Array(n);
    suf[n-1] = nums[n-1];
    for(let i = n - 2; i >= 0; i--) {
        suf[i] = suf[i + 1] | nums[i];
    }
    let l = 0, r = 0;
    let cur = nums[0];
    let freq = new Array(32).fill(0);
    for(let j = 0; j < 32; j++) {
        if((nums[0] & (1 << j)) !== 0) {
            freq[j]++;
        }
    }
    while (l < n) {
        while (r + 1 < n && cur !== suf[l]) {
            r++;
            cur |= nums[r];
            for(let j = 0; j < 32; j++) {
                if((nums[r] & (1 << j)) !== 0) {
                    freq[j]++;
                }
            }
        }
        res[l] = Math.max(1, r - l + 1);
        for(let j = 0; j < 32; j++) {
            if((nums[l] & (1 << j)) !== 0) {
                freq[j]--;
                if(freq[j] === 0) {
                    cur = cur & (~(1 << j));
                }
            }
        }
        l++;
    }
    return res;
};